Cached widget replay

AnOctopus opened the pull request:
## 📚 Context

Replay of cached elements, but for widgets

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- Convert memo and singleton to 2 stage caches that treat widget values as inputs
- Dynamically track widgets seen when executing a cached function, to allow the above to work
- Replay widget registration and messages from cached function results

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change


## 🧪 Testing Done

- [ ] Screenshots included
- [x] Added/Updated unit tests
- [x] Added/Updated e2e tests

AnOctopus committed:
Save initial implementation progress

Lots of internal docs writing that brought up a number of questions,
because I realized a small flaw in the original design and followed the
implications.

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Save WIP state to show others

AnOctopus committed:
Some simplification and test fixes

AnOctopus committed:
Extract widget key calculation

AnOctopus committed:
Move almost everything out of outer try block

AnOctopus committed:
Add basic widget replay test

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Add test for replay with dynamic widget lists

AnOctopus committed:
Add some missing type variables

AnOctopus committed:
Pipe cache_type through widget cache key generation

Not obviously worth it, but easily changed later if desired

AnOctopus committed:
Temp silence mypy about field access

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Fix import cycle

AnOctopus committed:
Mypy ignore memo import in e2e

No idea why it's this dumb

AnOctopus committed:
Make dynamic widget replay e2e grab the right thing

AnOctopus committed:
Access text via data-testid

AnOctopus committed:
Update seen widgets for cache after execution

AnOctopus committed:
Use widget key method in singleton code

AnOctopus committed:
Allow widget keys to generate when widgets are missing

Otherwise there may be exceptions from lookup failures when reconnecting
to a running app. We could instead catch the exception somewhere and
treat it as a cache miss, but this way seems more robust, because it
handles a conditionally created widget that is not currently being used
by allowing caching to still happen based on the other widgets.

AnOctopus committed:
Convert assertion into warning

It is tripping up the integration test job

AnOctopus committed:
Prevent hypothetical invalid metadata access attempt

tconkling commented on lib/streamlit/runtime/caching/__init__.py:
```diff
@@ -64,6 +65,11 @@ def save_block_message(
 
+def save_widget_metadata(metadata: WidgetMetadata[Any]) -> None:
+    MEMO_MESSAGES_CALL_STACK.save_widget_metadata(metadata)
+    SINGLETON_MESSAGE_CALL_STACK.save_widget_metadata(metadata)
```
Can you add a short docstring explaining why we do this and why it works? (At first I was like "how in the world is this safe", and then I looked at the CALL_STACK variables and saw that they are thread-locals.)

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -58,16 +64,24 @@
+class WidgetMsgMetadata:
+    widget_id: str
+    widget_value: Any
+    metadata: WidgetMetadata[Any]
```
Can you add a short docstring explaining the class's purpose?

tconkling commented on lib/streamlit/runtime/caching/memo_decorator.py:
```diff
@@ -476,19 +516,29 @@ def _read_from_mem_cache(self, key: str) -> bytes:
                 _LOGGER.debug("Memory cache MISS: %s", key)
                 raise CacheKeyNotFoundError("Key not found in mem cache")
 
+    def _read_initial_from_mem_cache(self, key: str) -> InitialCachedResults:
```
Short docstring? In particular, call out what the failure states are here (result not found _or_ result is wrong type), and what happens on failure (raise CacheKeyNotFoundError). 

(I don't think docstrings are necessary on every internal method, but the behavior is hard to infer from the function names here. "InitialCachedResults" vs "CachedResult" is slightly breaking my brain.)

tconkling commented on lib/streamlit/runtime/caching/singleton_decorator.py:
```diff
@@ -30,10 +31,13 @@
     CacheMessagesCallStack,
     CachedResult,
+    ElementMsgData,
+    InitialCachedResults,
```
I know this isn't new to this PR, but I wonder if there's a better name than "InitialCachedResults". "CachedDataAndReferencedWidgets"?

AnOctopus replied:
I'd like a better name too, though I'm not fond of "CachedDataAndReferencedWidgets". Is `PartialCachedResults` more intuitive to you?

tconkling replied:
Yes!

tconkling commented on lib/streamlit/runtime/state/safe_session_state.py:
```diff
@@ -123,3 +123,13 @@ def __delitem__(self, key: str) -> None:
 
             del self._state[key]
+
+    def get(self, key: str, default: Any = None) -> Any:
```
Needs a small unit test in safe_session_state_test.py

tconkling commented on lib/streamlit/runtime/state/session_state.py:
```diff
@@ -668,6 +668,12 @@ def get_stats(self) -> List[CacheStat]:
+        try:
+            return self[key]
+        except KeyError:
+            return default
```
Also needs a small unit test

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -186,10 +186,36 @@ def register_widget(
+def register_widget_from_metadata(
+    metadata: WidgetMetadata[T],
+    ctx: Optional["ScriptRunContext"],
+    # TODO get real values for these
```
What does this TODO mean? How do we know when it's been addressed?

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -186,10 +186,36 @@ def register_widget(
+    # TODO get real values for these
+    widget_func_name: Optional[str],
+    element_type: ElementType,
+) -> RegisterWidgetResult[T]:
```
Docstring plz :)

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -186,10 +186,36 @@ def register_widget(
+    widget_func_name: Optional[str],
+    element_type: ElementType,
+) -> RegisterWidgetResult[T]:
+    import streamlit.runtime.caching as caching
```
Is it important that this import be local? If yes, can we document it?

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -186,10 +186,36 @@ def register_widget(
+
+    widget_id = metadata.id
+    user_key = widget_id.split("-", maxsplit=2)[-1]
+    user_key = None if user_key == "None" else user_key
```
Can we pull this out into a helper function with its own docstring? (`get_widget_user_key` or whatever?) The comparison against the "None" literal - why does that happen? 

I think we need unit tests to enforce whatever the expected behavior is here, because on its own this feels brittle.

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -214,17 +240,7 @@ def register_widget(
-        callback_args=args,
-        callback_kwargs=kwargs,
-    )
+    caching.save_widget_metadata(metadata)
```
Can you add an inline comment saying why caching needs to know about this widget metadata?

tconkling commented on lib/tests/streamlit/runtime/caching/memo_test.py:
```diff
@@ -34,7 +39,24 @@
 
 
 def as_cached_result(value):
```
docstring plz (and also types, if we can get them? - but even `def as_cached_result(value: Any) -> Any` is helpful; as written, this is hard to parse as a new reader of the code)

tconkling commented on lib/tests/streamlit/runtime/caching/memo_test.py:
```diff
@@ -34,7 +39,24 @@
+    return initial
+
+
+def as_replay_test_data():
```
docstring + types plz

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -214,17 +240,7 @@ def register_widget(
-        callback_args=args,
-        callback_kwargs=kwargs,
-    )
+    caching.save_widget_metadata(metadata)
```
Also, I might have just overlooked it, but is this behavior tested anywhere?

AnOctopus replied:
Not by unit tests, but its role in making replay work is tested by some cypress tests.

tconkling reviewed (commented):
This looks thoughtful and thorough, BUT it also feels a bit write-only. That is: it's very complex, and it's hard for me (a new reader of the code), to find a way into that complexity, and to understand what's going on and why.

I have a couple general concerns
- Lots of methods are missing docstrings, and also don't have unit tests backing them up, so it's really hard to know "what is the contract this method is making and how do I know if I've broken it". Like, I don't feel confident I could make changes in here.
- Less critically, some of the naming here isn't helping the situation.

In between now and when this is ready to land, can we give any non-obvious, non-trivial method gets a docstring and enough unit tests to make their meaning clear to new readers? (I'm happy to help with this if you'd like to pair on it!)

AnOctopus commented:
About what I expected. I hadn't spent any time on documenting it since the initial high level overview. Thanks for checking over it, I'll let you know when it's ready for another round.

AnOctopus committed:
Add top level caching hook function docstrings

AnOctopus committed:
Clean up unused imports

AnOctopus committed:
Document widget key and metadata saving stuff

AnOctopus committed:
Document exceptions raised by memo lookup functions

AnOctopus committed:
Document and deduplicate `as_cached_result` in tests

AnOctopus committed:
Document new register widget function

AnOctopus committed:
Return key from extracted function

AnOctopus committed:
Unit tests for `user_key_from_widget_id`

AnOctopus committed:
Add `get` tests for session state

AnOctopus committed:
Document `WidgetMsgMetadata`

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -52,22 +57,33 @@
+
+    widget_id: str
+    widget_value: Any
+    metadata: WidgetMetadata[Any]
```
It looks like this doesn't need to be mutable? Can we make it a NamedTuple instead?

tconkling replied:
Similarly, `ElementMsgData`, `BlockMsgData`, `CachedResult`, and `InitialCachedResults` seem like they don't need to be mutable; assuming I'm not missing something, can we convert them as well?

AnOctopus replied:
Made some of them frozen, which I think is strictly better than converting to NamedTuple. Left `CachedResult` and `InitialCachedResult` as is, to indicate mutability, even though they have internal mutability (mutation of their fields) rather than external (mutating what their fields point to)

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -92,6 +151,35 @@ class CachedResult:
     sidebar_id: str
 
 
+@dataclass
+class WidgetStateNotFound:
+    pass
```
Why is this a `@dataclass`?

AnOctopus replied:
to make sure equality will work as intended

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -79,6 +95,49 @@ class BlockMsgData:
+The cache for a decorated function's results is split into two parts to enable
+handling widgets invoked by the function.
+
+Widgets act as implicit additional inputs to the cached function, so they should
+be used when deriving the cache key. However, we don't know what widgets are
+involved without first calling the function! So, we use the first execution
+of the function with a particular cache key to record what widgets are used,
+and use the current values of those widgets to derive a second cache key to
+look up the function execution's results. The combination of first and second
+cache keys act as one true cache key, just split up because the second part depends
+on the first.
+
```
~I'm confused by this - why are widgets additional inputs to the cached function? If we _don't_ consider widgets to be part of the function's cache key, what is the behavior we get?~ I understand now after we chatted in Slack! But I think it may be useful to include a small example - the one we chatted about, even - in this docstring to help future readers understand?

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -92,6 +151,35 @@ class CachedResult:
+    pass
+
+
+WIDGET_STATE_NOT_FOUND = WidgetStateNotFound()
```
Does this actually need a special class? If it's just a sentinel, could we do something like

```
WIDGET_STATE_NOT_FOUND = object()
```

and lose the `WidgetStateNotFound` class?

AnOctopus replied:
I wanted it to be something that will never be a valid widget state value. It would be weird for someone to have a widget return an empty object instance, but not impossible, while one returning this sentinel would require actively trying to fuck with our internals

tconkling commented on lib/tests/streamlit/runtime/caching/common_cache_test.py:
```diff
@@ -476,6 +494,33 @@ def foo(i):
+            return r
+
+        r = foo(1)
+        # foo(1)
```
Is this commented line supposed to be here?

AnOctopus replied:
Oh, nope. I removed that when I made widget registration work in replay.

tconkling commented on lib/tests/streamlit/runtime/caching/common_cache_test.py:
```diff
@@ -476,6 +494,33 @@ def foo(i):
+            if element.WhichOneof("type") == "radio"
+        ]
+
+        # assert values == [1, 1]
```
It looks like we're missing an assertion against the expected values here?

tconkling commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -186,10 +186,58 @@ def register_widget(
+
+
+def user_key_from_widget_id(wid: str) -> Optional[str]:
+    """Extract the user key used to generate a widget id, from that id.
+
+    Returns `None` instead of `"None"` if there was no user key,
+    for compatibility with the rest of the codebase, which represents it that way.
+
+    TODO This will not work correctly if the user actually provides "None"
+    as a key, but we can't avoid this kind of problem while storing the
+    string representation of the no-user-key sentinel as part of the widget id.
+    """
+    user_key = wid.split("-", maxsplit=2)[-1]
+    user_key = None if user_key == "None" else user_key
+    return user_key
```
Are these caveats expressed in unit tests? "This will not work correctly" - what will happen if the user passes "None"? Should we detect that? Is there a way around this limitation (maybe by generating our widget keys differently?)

AnOctopus replied:
We could get around it fully by making the key-present-or-not indicated by something like a wrapper type, or at least make it implausible that someone runs into it on accident by making the no-key sentinel something like "STREAMLIT_NO_USER_KEY" (although that contributes to user keys being annoying to read, which they already kinda are)

tconkling reviewed (commented):
Putting a pin in this review after raising some concerns about the design with @jrieke, but I'll circle back!

AnOctopus committed:
Freeze some dataclasses

AnOctopus committed:
Remove unit test that doesn't test anything

AnOctopus committed:
Clarify `get_widget_user_key` behavior

AnOctopus commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -319,8 +413,10 @@ def maybe_show_cached_st_function_warning(
 
         """
-        if st_func_name in NONWIDGET_ELEMENTS:
+        # TODO does this leave anything unsupported? can we remove the warning callstack?
```
Some things are neither, replace this with a comment explaining that

AnOctopus committed:
Rearrange memo write ctx code as early return

AnOctopus committed:
Remove outdated comment

AnOctopus committed:
Update comment about widget vs nonwidget for cache warning callstack

AnOctopus committed:
Add code example to `Cache result structure` note

AnOctopus committed:
Simplify code by ignoring missing widgets for cache key

AnOctopus committed:
Implement contains for session state

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Add `experimental_allow_widgets` arg to memo/singleton

AnOctopus committed:
Unparenthesize context manager stack

AnOctopus committed:
Allow widgets in relevant cache tests

AnOctopus committed:
Actually pass `allow_widgets` to cache instance

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Remove unused mypy bypass

tconkling commented on e2e/scripts/widget_replay_dynamic.py:
```diff
@@ -0,0 +1,32 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Can you add a short docstring at the top of the test explaining what it's testing? (What is it doing differently from `widget_replay`?)

tconkling commented on e2e/scripts/widget_replay.py:
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
```
Nit: can we rename this (and the following test) to `cached_widget_replay` (or memoized_widget_replay or whatever) for clarity?

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -80,6 +96,62 @@ class BlockMsgData:
 
 MsgData = Union[ElementMsgData, BlockMsgData]
 
+"""
+Note [Cache result structure]
+
+The cache for a decorated function's results is split into two parts to enable
+handling widgets invoked by the function.
+
+Widgets act as implicit additional inputs to the cached function, so they should
+be used when deriving the cache key. However, we don't know what widgets are
+involved without first calling the function! So, we use the first execution
+of the function with a particular cache key to record what widgets are used,
+and use the current values of those widgets to derive a second cache key to
+look up the function execution's results. The combination of first and second
+cache keys act as one true cache key, just split up because the second part depends
+on the first.
+
```
This is great, thanks for the detailed docs!

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -239,14 +349,15 @@ def get_or_create_cached_value():
-                            return_value = func(*args, **kwargs)
-                    else:
-                        return_value = func(*args, **kwargs)
+                with cached_func.warning_call_stack.calling_cached_function(func):
+                    with cached_func.message_call_stack.calling_cached_function():
+                        with cached_func.message_call_stack.maybe_allow_widgets(
+                            cached_func.allow_widgets
+                        ):
+                            with cached_func.warning_call_stack.maybe_suppress_cached_st_function_warning(
+                                cached_func.suppress_st_warning
+                            ):
+                                return_value = func(*args, **kwargs)
```
Nit: can these nested `with` blocks be combined?

```python
with (
    cached_func.warning_call_stack.calling_cached_function(func),
    cached_func.message_call_stack.calling_cached_function(),
    cached_func.message_call_stack.maybe_allow_widgets(cached_func.allow_widgets),
    cached_func.warning_call_stack.maybe_suppress_cached_st_function_warning(cached_func.suppress_st_warning)
):
    return_value = func(*args, **kwargs)
```

AnOctopus replied:
Nope, that's a python 3.8 (I think, might be 9) feature. That's how I originally wrote it, but it caused the 3.7 CI job to fail

tconkling replied:
I think it may just be the parens that are python 3.9+? Earlier Pythons definitely support multiple context managers per `with` (we use the pattern in some of our tests, with some ugly line wrapping). (But it may be ugly enough to not be worthwhile!)

AnOctopus replied:
Oh, right. I tried that before the parenthesized version and totally forgot about it. Something about python's parser requires the linebreaks to be in the middle of each expression, or else it won't parse, and the result is pretty bad.

```py
with cached_func.warning_call_stack.calling_cached_function(
func
), cached_func.message_call_stack.calling_cached_function(
), cached_func.message_call_stack.maybe_allow_widgets(
cached_func.allow_widgets
), cached_func.warning_call_stack.maybe_suppress_cached_st_function_warning(
cached_func.suppress_st_warning
):
    return_value = func(*args, **kwargs)
```
Something like this, though probably with slightly different indentation.

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -175,15 +275,25 @@ def replay_result_messages(
+    ctx = get_script_run_ctx()
 
     try:
         for msg in result.messages:
             if isinstance(msg, ElementMsgData):
+                if msg.widget_metadata is not None:
+                    register_widget_from_metadata(
+                        # TODO: can we provide real values for these?
+                        msg.widget_metadata.metadata,
+                        ctx,
+                        "",
+                        "",
+                    )
```
Is this TODO important? What are the ramifications of not having real values?

AnOctopus replied:
IIRC the values are used for generating the `DuplicateWidgetID` error message, so not having them means the error message is worse

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -417,9 +543,31 @@ def save_element_message(
         """
         id_to_save = self.select_dg_to_save(invoked_dg_id, used_dg_id)
         for msgs in self._cached_message_stack:
-            msgs.append(
-                ElementMsgData(delta_type, element_proto, id_to_save, returned_dg_id)
-            )
+            try:
+                wid = element_proto.id  # type: ignore
+                # TODO replace `Message` with a more precise type
+                if not self._registered_metadata:
+                    _LOGGER.warning(
+                        "Trying to save widget message that wasn't registered. This should not happen."
+                    )
+                    raise AttributeError
```
Can you add some details about why this might happen? Is it just a sanity check and it literally means "something went terribly wrong"? If so, should we consider raising an error that won't be caught right below, because the program is probably in an undefined state? (Basically, is this truly a recoverable error?)

AnOctopus replied:
If streamlit is running, it should never happen, and indicates something is broken. But, if streamlit isn't running the script, widget registration never happens, yet we can still end up in that caching code. The current behavior here was a hack to get integration tests working. Though now that I'm thinking about it again, I could check if the runtime exists, so it can silently continue or explode, as appropriate.

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -417,9 +543,31 @@ def save_element_message(
-                ElementMsgData(delta_type, element_proto, id_to_save, returned_dg_id)
-            )
+            try:
+                wid = element_proto.id  # type: ignore
```
~Why `# type: ignore`?~ Edit: oh, I see - it's because it's just the protobuf message base class.

Could we define and use a `Protocol` here to achieve type safety?

tconkling commented on lib/streamlit/runtime/caching/memo_decorator.py:
```diff
@@ -453,11 +475,44 @@ def write_result(self, key: str, value: Any, messages: List[MsgData]) -> None:
         """Write a value and associated messages to the cache.
         The value must be pickleable.
         """
+        main_id = st._main.id
+        sidebar_id = st.sidebar.id
```
Micronit: can we move this logic below the null-ctx fast fail exit?

tconkling commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -93,6 +165,32 @@ class CachedResult:
+    """
+
+    widget_ids: Set[str]
+    results: Dict[str, CachedResult]
```
`results` is not an intuitive name to me - I know it's more verbose, but what about something like `results_by_widgets_hash` instead?

I also wonder if this class could have a more descriptive name as well  - it's not really "initial cached results", it's more "set of all thus-far computed results for a given function". What about `CachedResultSet` or something? `CachedResultContainer`?

AnOctopus replied:
@tconkling 
What do you think about my previous idea of `PartialCachedResults`? Or how about something like `ArgKeyCachedResult`, to distinguish it from the full results that incorporate the widget state derived key?

tconkling commented on lib/tests/streamlit/runtime/caching/common_cache_test.py:
```diff
@@ -46,6 +52,18 @@ def get_text_or_block(delta):
+    """
+    result = CachedResult(value, [], st._main.id, st.sidebar.id)
+    widget_key = _make_widget_key([], cache_type)
+    d = {}
+    d[widget_key] = result
```
Nit: `d = {widget_key: result}`

tconkling commented on lib/tests/streamlit/runtime/caching/memo_test.py:
```diff
@@ -18,25 +18,49 @@
+    return _as_cached_result(value, CacheType.MEMO)
+
+
+def as_replay_test_data():
```
Can you add the return type?

tconkling reviewed (changes requested):
Code is much better documented and easier to follow - thanks! Feels like the big missing thing here is thorough tests (probably in `common_cache_test.py`):

- The experimental config option works as expected
- Calling a widget-replay function with an unseen widget key works (computes new cache value)
- Calling a widget-replay function with an already-seen widget key works (does not compute new cache value)
- Mutating a widget inside a function generates the expected widget key

AnOctopus committed:
Move memo code body below early exit

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Small cache test cleanup

AnOctopus committed:
Type annotation

AnOctopus committed:
Clearer names for cached widget replay e2e tests

AnOctopus committed:
Add explanation of purpose to dynamic widget replay test script

AnOctopus commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -349,7 +470,10 @@ def maybe_show_cached_st_function_warning(
+        # There are some elements not in either list, which we still want to warn about.
+        # Ideally we will fix this by either updating the lists or creating a better
+        # way of categorizing elements.
+        if st_func_name in NONWIDGET_ELEMENTS or st_func_name in WIDGETS:
```
Just noticed that I still need to update this to only skip warning about widgets if they are allowed

AnOctopus committed:
Slightly more mypy-friendly checking of element id

AnOctopus committed:
Warn on widget only when they are disabled

The cache contextmanager stack of doom grows

AnOctopus committed:
Fix message stack naming inconsistency

AnOctopus committed:
Test widget key cache hit/miss for widget replay

AnOctopus committed:
Rename InitialCachedResults -> MultiCacheResults

Initial is misleading, and while Multi is a bit vague, that is
preferable. Coming up with a good name for this is hard.

AnOctopus committed:
Update method/variable names to reflect rename

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Use text instead of write for reliable output formatting

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

tconkling commented on e2e/specs/cached_widget_replay.spec.js:
```diff
@@ -0,0 +1,34 @@
+  it("replays widget on rerun", () => {
+    cy.get(".stRadio").should("have.length", 1);
+    cy.get(".stButton").first().click();
+    cy.get(".stRadio").should("have.length", 1);
```
Not positive, but I think this might not accurately test what we're looking for? Since `cy.get(".stRadio").should("have.length", 1)` is true before the button click, it will almost be true immediately after the button click, because the big DOM update won't happen until after the script is done re-running.

Can we insert a "wait for a complete rerun" bit between the button click and the second number-of-radio-buttons check? (Maybe: "check that the rerun indicator to exists" followed by "check for the rerun indicator to not exist"?)

tconkling commented on e2e/specs/cached_widget_replay_widget_key.spec.js:
```diff
@@ -0,0 +1,36 @@
+  it("runs cached function with new widget values", () => {
+    cy.get(".stRadio").should("have.length", 1);
+    cy.get("[data-testid='stText']").should("have.text", "['function ran']");
+    cy.get(".stButton").first.click();
```
This (and the button click below) probably also need an explicit "wait for rerun" step (assuming I'm not mistaken about the need for it.)

tconkling commented on lib/streamlit/runtime/caching/memo_decorator.py:
```diff
@@ -453,11 +475,45 @@ def write_result(self, key: str, value: Any, messages: List[MsgData]) -> None:
+        else:
+            widgets = set()
+
+        first_stage_results: Optional[MultiCacheResults] = None
```
Nit: `first_stage_results` implies that a second data structure will hold the results from the "second stage" of cache lookup, but I don't think this is the case, right? (It's just that the results we're looking for may or may not yet be in the `MultiCacheResults` - and if they're not, we'll update that `MultiCacheResults` with the results when we have them.)

How about just `multi_cache_results` instead?

tconkling commented on lib/tests/streamlit/runtime/caching/memo_test.py:
```diff
@@ -18,25 +18,49 @@
-    return CachedResult(value, [], st._main.id, st.sidebar.id)
+
+def as_replay_test_data() -> MultiCacheResults:
+    """Creates cached results for a function that returned `value`
```
What does `value` refer to in "a function that returned `value`"? (Is this docstring maybe intended for a different function?)

tconkling reviewed (approved):
Looks good! I had a hard time determining this from reading the code, so I want to verify some behavior:

Let's say we have a cached function with disk persistence enabled. We get a `MultiCacheResults` hit for our function params, but our particular `CachedResult` isn't yet computed. After it's computed and we update `MultiCacheResults`, will that updated results object be written back to disk? (In other words: do we properly write mutated `MultiCacheResults` back out to disk?)

AnOctopus committed:
Wait for rerun indicator to come and go in tests

AnOctopus committed:
Give up and use sleep instead

AnOctopus committed:
Try long sleeps to see if that makes the tests pass

AnOctopus committed:
Update docstring

AnOctopus committed:
Rename local cache results variable

AnOctopus committed:
Actually call click function

Typechecking plz save me

AnOctopus committed:
Does a shorter sleep let it not stop in the middle?

AnOctopus committed:
The cypress name for "sleep" is "wait"

AnOctopus committed:
Merge branch 'develop' into feature/cached-widget-replay

AnOctopus committed:
Restore waiting for rerun indicator

AnOctopus committed:
More specific button selection

AnOctopus committed:
Extract waiting for rerun into a function

AnOctopus committed:
Use cached delta_type in duplicate widget error message
